driver/src/main/java/org/neo4j/driver/internal/handlers/PullAllResponseHandler.java

```diff
@@ -102,56 +104,56 @@ public synchronized void onRecord( Value[] fields )
     {
         Record record = new InternalRecord( runResponseHandler.statementKeys(), fields );
 
-        if ( recordPromise != null )
+        if ( recordFuture != null )
         {
-            succeedRecordPromise( record );
+            completeRecordFuture( record );
         }
         else
         {
             queueRecord( record );
         }
     }
 
-    public synchronized InternalFuture<Record> nextAsync()
+    public synchronized CompletionStage<Record> nextAsync()
```

Comment:
I think we can improve concurrent performance among `#onRecord()` and `#nextAsync()` methods by avoiding method level  `synchronized` access. We will have to keep an eye on `recordFuture` and queuing records (we can maybe use `ConcurrentLinkedQueue` instead of `Queue`).